date: add new date class

sousajo-cc opened the pull request.

Commit:
date: add new date class

Commit:
more tests for date creation

Commit:
date to string

Commit:
space ship operator *chef kiss*

Commit:
very bad wrappers for year, month and day

Commit:
adding year but perhaps in the wrong way

soaresiv committed:
Add operator+ on Date

Commit:
date: add logic to add months to a date

Commit:
date: start day adding logic

Commit:
add days test

Commit:
printable expected

Commit:
printable expected

Commit:
test fix

Commit:
making error printable

Commit:
add day logic finished

Commit:
extracted concept

Commit:
moved free variable into function

Commit:
making structs into proper classes

Commit:
const, constexpr and nodiscard

Commit:
more small stuff

Commit:
arithmetic for day, month and year spans

Commit:
ran clang formater

Commit:
date: getWeekDay implemented.

Commit:
inline!

Commit:
more operators

Commit:
week duration

Commit:
concept for duration

Commit:
don't repeat yourself, I guess

Commit:
more tests

Commit:
date: simplifying(?) duration concept definition

Commit:
util: refactor map

excuse to use ranges and template abbreviation with constrained
auto.

luizchagasjardim commented on src/imgui/date.hpp:
```diff
@@ -42,41 +42,41 @@ inline std::string toString(DaysOfTheWeek d)
+  [[nodiscard]] constexpr explicit Day(int v) : durationValue{ v } {}
   [[nodiscard]] constexpr auto operator<=>(Day const &) const = default;
-  int value{};
+  int durationValue{};
```
:+1:

luizchagasjardim commented on src/imgui/date.hpp:
```diff
@@ -99,50 +99,35 @@ struct Month
+concept Duration = requires(T c)
+{
+  c.durationValue;
+};
```
This change I don't like, because one day there may be another struct somewhere with a field called durationValue, and then this could cause a Weird compilation error. If you want to simplify this concept, we could make a base class for all of them and make the concept be "is derived from this class".

sousajf1 replied:
I agree yes. I changed the field name because of that but indeed is not a strong guarantee. maybe going the inheritance  way is the way. I guess after that this is ready to review

luizchagasjardim replied:
We're still missing some things that I think would be nice:
operator-(Date, Date) -> Day
separating duration from date value
month.toString()
putting everything in namespaces

sousajf1 replied:
cool

luizchagasjardim commented on src/imgui/util.hpp:
```diff
@@ -3,21 +3,21 @@
-  output.resize(input.size());
-  std::transform(input.begin(), input.end(), output.begin(), f);
+  std::vector<Out> output{};
+  for (auto val : input | std::ranges::views::transform(f)) {
```
Very cool :+1:

Commit:
date: added getMonthName() and refactored Duration concept

Commit:
date: put everything into namepace

there was a namespace lookup bug on gcc we worked around it
without knowing what we were doing.

Commit:
days until tests

Commit:
unary operator- for duration structs

Commit:
de primeira kkkk

Commit:
appeased clang tidy

Commit:
remove using from header

luizchagasjardim commented:
Still missing separating durations from values and refactoring tests.

Commit:
date: appease clang-format

Commit:
date: template abbreviation with constrained auto

soaresiv reviewed (approved):
Looks good.
For those who want to learn C++ I guess this PR is not a good example to take.

luizchagasjardim commented:
> 
> 
> Looks good. For those who want to learn C++ I guess this PR is not a good example to take.

Agreed, but most of the more advanced stuff was added with a lot of people in the meeting.

sousajf1 commented:
@all I'll merge this

Commit:
Merge branch 'main' into sj_datetime_evryopne_happy
